### PR TITLE
Network : remove unused residual tower weight values

### PR DESCRIPTION
**Pull request by ihavnoid**
We restructured the code ad moved everything to ForwardPipe.  No need to leave it in the memory anymore.  This saves about half a gigabyte of memory footprint on 20x256 nets

### CODE REVIEW

**Comment by ihavnoid** [comment]
Just to be clear, there is no guarantee that the memory is actually saved because it has a lot to do with how  libc manages memory.  If it refuses to return memory back to OS, then it will still eat up memory footprint.   But at least this is a workable, simple change. :)

File: src/CPUPipe.cpp [hunk 1]
```diff
@@ -348,6 +348,24 @@ void CPUPipe::push_input_convolution(unsigned int /*filter_size*/,
...
+void CPUPipe::move_residual(unsigned int /*filter_size*/,
+                            unsigned int /*channels*/,
+                            unsigned int /*outputs*/,
+                            std::vector<float>&& weights_1,
+                            std::vector<float>&& means_1,
+                            std::vector<float>&& variances_1,
+                            std::vector<float>&& weights_2,
+                            std::vector<float>&& means_2,
+                            std::vector<float>&& variances_2) {
+    m_conv_weights.push_back(std::move(weights_1));
```
Comment (gcp) [inline src/CPUPipe.cpp]: This is the case where you *really* want emplace_back, isn't it? So you get a move constructor, not a copy constructor...?

Comment (gcp) [reply src/CPUPipe.cpp]: Also fix the input convolution?

File: src/ForwardPipe.h [hunk 2]
```diff
@@ -49,6 +49,22 @@ class ForwardPipe {
                                const std::vector<float>& means_2,
                                const std::vector<float>& variances_2) = 0;
 
+    // have a better way to deal with situations when caller doesn't need the prior net?
+    // implement this.
+    virtual void move_residual(unsigned int filter_size,
```
Comment (gcp) [inline src/ForwardPipe.h]: Can't you support both cases by the combination of emplace_back and the universal references?

File: src/Network.cpp [hunk 3]
```diff
@@ -499,19 +509,31 @@ void Network::initialize(int playouts, const std::string & weightsfile) {
...
     if (use_selfcheck) {
-        m_forward_cpu = init_net(std::make_unique<CPUPipe>());
+        m_forward_cpu = init_net(std::make_unique<CPUPipe>(), true);
     }
 #else
     (void)use_selfcheck;
 #endif
+
+    // ForwardPipe instances are initialized.  We don't need the residual tower nets anymore
+    // clear() doesn't guarantee memory freeing, so this is an alternative...
```
Comment (gcp) [inline src/Network.cpp]: Use clear() + shrink_to_fit().

**Comment by gcp** [comment]
Apparently another trick is to make the functions take the vectors by value(!) and then emplace_back(std::move(value_arg)) into the members.

This will copy in normal cases, and move when the vector argument is std::moved from the callers' site.

It makes no sense to me but then again so does most of C++ these days.

File: src/CPUPipe.cpp (hunk 1 above)
Comment (ihavnoid) [reply src/CPUPipe.cpp]: emplace_back(std::move(…)) works.
emplace_back(…) still consumes peak memory footprint.

Also fixed convolution input but the gains should be much smaller.

File: src/Network.cpp (hunk 3 above)
Comment (ihavnoid) [reply src/Network.cpp]: Got it.

**Comment by ihavnoid** [comment]
Hmm... call by value + std::move might work.  I will just rewrite it that way as that seems to be the most c++14-ish way.

Makes no sense (haha) but it seems that will work. :)

**Comment by ihavnoid** [comment]
Thinking a bit more, I think I was trying to solve the problem using a solution that is too complicated.  Pushed a change that simply stores the pointer of the weights on CPUPipe.

**Comment by gcp** [comment]
Seems like we'd want a separate network-data class and pass shared ptrs to that?

**Comment by ihavnoid** [comment]
3rd attempt pushed.  This time everything is not encapsulated into a ForwardPipeWeights subclass, and shared_ptr<const ForwardPipeWeights> is passed around.

**Pull request merged**